source/mir/random/generic.d

```diff
@@ -0,0 +1,384 @@
+        // t_m is t_l or t_r wherever larger f(x) larger
+        if (bly > bry)
+            t_m = t_l;
+        else
+            t_m = t_r;
+
+        LinearFun!S hat;
+        LinearFun!S squeeze;
+
+        with(T1Type)
+        final switch(t1)
+        {
+            // concave near b_l and t_r(x) <= f(x) <= t_l(x)
+            case T1a:
+                hat = t_l;
+                squeeze = t_r;
+                break;
+
+            // convex near b_l and t_l(x) <= f(x) <= t_r(x)
+            case T1b:
+                hat = t_r;
+                squeeze = t_l;
+                break;
+
+            // concave near b_l and r(x) <= f(x) <= t_l(x)
+            case T2a:
+                hat = t_l;
+                squeeze = r;
+                break;
+
+            // convex near b_l and r(x) <= f(x) <= t_r(x)
+            case T2b:
+                hat = t_r;
+                squeeze = r;
+                break;
+
+            // concave near b_l and t_r(x) <= f(x) <= r(x)
+            case T3a:
+                hat = r;
+                squeeze = t_r;
+                break;
+
+            // convex near b_l and t_l(x) <= f(x) <= r(x)
+            case T3b:
+                hat = r;
+                squeeze = t_l;
+                break;
+
+            // concave on [b_l, b_r] and r(x) <= f(x) <= t_m(x)
+            case T4a:
+                hat = t_m;
+                squeeze = r;
+                break;
+
+            // convex on [b_l, b_r] and t_m(x) <= f(x) <= r(x)
+            case T4b:
+                hat = r;
+                squeeze = t_m;
+                break;
+        }
+        return SqueezeAndHat!S(hat, squeeze);
+    }
+}
+
+unittest
+{
+    alias f0 = (x) => x ^^ 3;
+    alias f1 = (x) => 3 * x ^^ 2;
+    alias f2 = (x) => 6 * x;
+    alias ti = tinflexInternal!(f0, f1, f2);
+
+    with(T1Type)
+    {
+        // concave
+        assert(ti.determineType(-3, -1) == T4a);
+        // inflection point at x = 0, concave before
+        assert(ti.determineType(-1, 1) == T1a);
+        // convex
+        assert(ti.determineType(1, 3) == T4b);
+    }
+}
+
+unittest
+{
+    alias f0 = (x) => x ^^ 4;
+    alias f1 = (x) => 4 * x ^^ 3;
+    alias f2 = (x) => 12 * x * x;
+    alias ti = tinflexInternal!(f0, f1, f2);
+
+    with(T1Type)
+    {
+        // entirely convex
+        assert(ti.determineType(-3, -1) == T4b);
+        assert(ti.determineType(-1, 1) == T4b);
+        assert(ti.determineType(1, 3) == T4b);
+    }
+}
+
+unittest
+{
+    import std.math: cos, sin, PI;
+    alias f0 = (x) => sin(x);
+    alias f1 = (x) => cos(x);
+    alias f2 = (x) => -sin(x);
+    alias ti = tinflexInternal!(f0, f1, f2);
+
+    with(T1Type)
+    {
+        // type 1a: concave
+        assert(ti.determineType(0, 2 * PI) == T1a);
+        assert(ti.determineType(2 * PI, 4 * PI) == T1a);
+        assert(ti.determineType(2., 4) == T1a);
+        assert(ti.determineType(0., 5) == T1a);
+        assert(ti.determineType(1., 5) == T1a);
```

Comment:
Stylistic suggestion: always use an explicit `n.0` for floating-point, rather than this notation.  The characters/keystrokes saved are not worth it for the cost in readability/clarity of meaning.